src/pipes/input-characters-validation.ts

```diff
@@ -0,0 +1,40 @@
+import { BadRequestException, Injectable, InternalServerErrorException, PipeTransform } from '@nestjs/common';
+import { ACBS } from '@ukef/constants';
+
+@Injectable()
+export class InputCharacterValidationPipe implements PipeTransform {
+  transform(value: any, metadata) {
+    if (metadata.type === 'query' || metadata.type === 'param') {
+      this.recursiveCheck(
+        value,
+        ACBS.URL_ALLOWED_CHARACTERS_REGEX,
+        (key, invalidCharacters) => `URL field ${key} has invalid characters ${invalidCharacters}.`,
+        metadata.data,
+      );
+    } else if (metadata.type === 'body') {
+      this.recursiveCheck(
+        value,
+        ACBS.ALLOWED_CHARACTERS_REGEX,
+        (key, invalidCharacters) => `Field ${key} has invalid characters ${invalidCharacters}.`,
+        metadata.data,
+      );
+    } else {
+      throw new InternalServerErrorException(`Unknown input type '${metadata.type}'.`);
+    }
+
+    return value;
+  }
+
+  recursiveCheck(value, findCharactersRegex, errorMessageGenerator, key = null) {
+    if (typeof value === 'object' && value !== null) {
+      Object.keys(value).forEach((key) => {
+        this.recursiveCheck(value[`${key}`], findCharactersRegex, errorMessageGenerator, key);
```

Comment:
If we type `key` as type string then we probably can use `key` instead of  \`${key}\`

Reply:
It didn't like this :( I've changed it to use this instead
```
Object.entries(value).forEach(([k, v]) => {
        this.recursiveCheck(v, findCharactersRegex, errorMessageGenerator, k);
      });
```

Reply:
Ooh, I've realised that I made a typo when I tried this originally which is why it didn't work. Is there a preference of which to use?

Reply:
Neither [Object.keys()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys) or [Object.entries()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) enumerate properties in the prototype chain as well